Ensure SubscriptionPeriods are represented as 1week instead of 7days

fire-at-will opened the pull request:
### Checklist
- [x] If applicable, unit tests
- [x] If applicable, create follow-up issues for `purchases-android` and hybrids (N/A)

### Motivation
This PR resolves [CF-659](https://revenuecats.atlassian.net/browse/CF-659) & `react-native-purchases` issue [#348](https://github.com/RevenueCat/react-native-purchases/issues/348).

Occasionally, StoreKit will state that a period is 7 days instead of 1 week. This unexpected behavior can cause unintended issues for customers on their paywalls and generally decreases developer confidence in the Offerings/Products provided by the SDK.

### Description
- Implemented a `normalization` function to be run on the input parameters of the `SubscriptionPeriod` class that, if possible to do so evenly, converts days into weeks and months into years.
- This function is static since it needs to be executed before the `SubscriptionPeriod` class has completed instantiation.
- Conversion of weeks to months and weeks to years was intentionally left out since those conversion rates don't use whole numbers (variable number of weeks in a month and 52.14 weeks per year)

vegaro commented on Sources/Purchasing/StoreKitAbstractions/StoreKitWorkarounds.swift:
```diff
@@ -75,3 +75,34 @@ extension SK1Product {
+    /// of months into years if there are no leftover units after the conversion.
+    ///
+    /// Occassionally, StoreKit seems to send back a value 7 days for a 7day trial
+    /// instesad of a value of 1 week for a trial of 7 days in length.
```
Typo instead

vegaro commented on Sources/Purchasing/StoreKitAbstractions/SubscriptionPeriod.swift:
```diff
@@ -29,8 +29,7 @@ public class SubscriptionPeriod: NSObject {
 
-        self.value = value
-        self.unit = unit
+        (self.value, self.unit) = Self.normalizeValueAndUnits(value: value, unit: unit)
```
Love that destructuring :chefskiss:

vegaro reviewed (approved):
Looks good! Did you manage to reproduce it by the way? We should probably file a radar if you haven't done yet

NachoSoto commented on Tests/UnitTests/Purchasing/StoreKitAbstractions/StoreKitWorkaroundsTests.swift:
```diff
@@ -0,0 +1,62 @@
+            (112, .day, 16, .week),
+            (356, .day, 356, .day),
+            (712, .day, 712, .day),
+            // Test week simplicfication
```
```suggestion
            // Test week simplification
```

NachoSoto reviewed (changes requested):
Looks good in general, but I wouldn't normalize all `SubscriptionPeriod` creation, that feels like spooky action at a distance.
And because we don't have tests for `SubscriptionPeriod` initialization (since it was just a basic data type), this is accidentally mutating the data that goes in.

Instead of `normalize` though, since we want to keep the type immutable, I would change it to `normalized`:
```swift
extension SubscriptionPeriod { 
  func normalized() -> Self {
    return .init(...)
  }
}
```

fire-at-will commented on Sources/Purchasing/StoreKitAbstractions/StoreKitWorkarounds.swift:
```diff
@@ -75,3 +75,34 @@ extension SK1Product {
+    ) -> (value: Int, unit: Unit) {
+        switch unit {
+        case .day:
+            if value % 7 == 0 {
```
Whaaaaat I didn't know Integer had this function!! 🤯🤯

fire-at-will commented on Sources/Purchasing/StoreKitAbstractions/StoreKitWorkarounds.swift:
```diff
@@ -75,3 +75,34 @@ extension SK1Product {
+    internal static func normalizeValueAndUnits(
+        value: Int,
+        unit: Unit
+    ) -> (value: Int, unit: Unit) {
```
I don't think making this a mutating function on `SubscriptionPeriod` is possible since both `value` and `unit` are `let` constants. Additionally, I'm not sure if making this function's signature
```swift
internal static func normalize(_ subscriptionPeriod: SubscriptionPeriod) -> SubscriptionPeriod
```
is feasible either since the `normalize` function is called from within `SubscriptionPeriod`'s initializer, and such a call would lead to an endless call loop.

I'm not a fan of the function being static, but I think the only alternative is having the developer call `normalize` every time a `SubscriptionPeriod` is instantiated, which doesn't seem like a good design.

fire-at-will replied:
Oh, I just saw [this comment](https://github.com/RevenueCat/purchases-ios/pull/1591#discussion_r878358928) 😅

I think it makes a lot of sense! I'll adopt that approach since this is just meant to be a StoreKit fix.

fire-at-will commented on Tests/UnitTests/Purchasing/StoreKitAbstractions/StoreKitWorkaroundsTests.swift:
```diff
@@ -0,0 +1,62 @@
+    func testSimplifyValueAndUnits() {
+        let expectations: [(
+            inputValue: Int, inputUnit: SubscriptionPeriod.Unit,
+            expectedValue: Int, expectedUnit: SubscriptionPeriod.Unit
```
Great feedback! It's so convenient that Nimble has the `description` parameter 🙂

NachoSoto commented on Sources/Purchasing/StoreKitAbstractions/StoreKitWorkarounds.swift:
```diff
@@ -75,3 +75,31 @@ extension SK1Product {
+        case .day:
+            if value.isMultiple(of: 7) {
+                let numberOfWeeks = value / 7
+                return SubscriptionPeriod(value: numberOfWeeks, unit: .week)
```
This could also be:
```suggestion
                return .init(value: numberOfWeeks, unit: .week)
```

NachoSoto commented on Sources/Purchasing/StoreKitAbstractions/StoreKitWorkarounds.swift:
```diff
@@ -75,3 +75,31 @@ extension SK1Product {
+            break
+        }
+
+        return SubscriptionPeriod(value: value, unit: unit)
```
Now you can just do this 😄 
```suggestion
        return self
```

fire-at-will replied:
You're right! 🤯 
The fewer objects the SDK creates in RAM, the better 😃

NachoSoto commented on Tests/UnitTests/Purchasing/StoreKitAbstractions/StoreKitWorkaroundsTests.swift:
```diff
@@ -0,0 +1,69 @@
+        ]
+
+        for expectation in expectations {
+            let outputPeriod = SubscriptionPeriod(
+                value: expectation.inputValue,
+                unit: expectation.inputUnit
+            ).normalized()
```
Maybe it makes sense to compare the whole period now?
```swift
let period = SubscriptionPeriod(
  value: expectation.inputValue,
  unit: expectation.inputUnit
)
let normalized = period.normalized()
let expected = SubscriptionPeriod(
  value: expectation.expectedValue,
  unit: expectation.expectedUnit
)

expect(period).to(
  equal(expected),
  description: "..."
)
```

fire-at-will replied:
That's a great way to clean this test up, made the changes

NachoSoto commented on Sources/Purchasing/StoreKitAbstractions/SubscriptionPeriod.swift:
```diff
@@ -64,6 +65,7 @@ public class SubscriptionPeriod: NSObject {
         }
 
         return .init(value: sk2SubscriptionPeriod.value, unit: unit)
+            .normalized()
```
Awesome.

fire-at-will commented:
@NachoSoto @vegaro Thanks for the feedback! 😄 

I made the following changes in [this commit](https://github.com/RevenueCat/purchases-ios/pull/1591/commits/3d2af2fb7b3147f1f49fe373ee73422825bc3bdc):
- Change the `normalizeValueAndUnit` function signature to `internal func normalized() -> SubscriptionPeriod`. Note that this is no longer `static` 🥳
- Move `normalized` call site from `SubscriptionPeriod.init` to `SubscriptionPeriod.from(sk1SubscriptionPeriod: SKProductSubscriptionPeriod)` and `SubscriptionPeriod.from(sk2SubscriptionPeriod: StoreKit.Product.SubscriptionPeriod)` so the fix is only used when the `SubscriptionPeriod` is created from StoreKit.
- Added descriptions to the assertions in the unit test.
- Change normalization logic to use `Int.isMultipleOf` instead of the modulo operator.
- Fixed typos 😎

NachoSoto commented on Sources/Purchasing/StoreKitAbstractions/SubscriptionPeriod.swift:
```diff
@@ -55,6 +55,7 @@ public class SubscriptionPeriod: NSObject {
         }
 
         return .init(value: sk1SubscriptionPeriod.numberOfUnits, unit: unit)
+            .normalized()
```
We can't test SK2's because it has no initializers, but maybe add a test in `SubscriptionPeriodTests` that ensures this is called here? It doesn't have to be comprehensive at all since you have all those examples in `StoreKitWorkaroundsTests`, maybe just one.

fire-at-will replied:
Oooh, added in one with a few different scenarios to test. Great idea!

NachoSoto reviewed (approved):
🎉

NachoSoto commented:
I'm excited for this change.
I just realized I had this in my app:
```swift
var periodTitle: String {
	assert(self.unit == .week, "Only week trials supported: \(self.unit)")
	
	return Plurals.Pro.Trial.intervalWeeks(self.value)
}
```
The `assert` won't run in production, but I'd probably be displaying "7 weeks" if users ran into this issue 😵‍💫

NachoSoto commented on Tests/UnitTests/Purchasing/StoreKitAbstractions/SubscriptionPeriodTests.swift:
```diff
@@ -73,4 +73,63 @@ class SubscriptionPeriodTests: TestCase {
+    /// Necessary since SKProductSubscriptionPeriod.debugDescription & SKProductSubscriptionPeriod.description
+    /// return the object's address in memory.
+    @available(iOS 11.2, *)
+    private func description(for skProductSubscriptionPeriod: SKProductSubscriptionPeriod) -> String {
```
Instead of doing this, let's implement `CustomStringConvertible ` in the tests:
```swift
extension SubscriptionPeriod: CustomStringConvertible {

  var description: String {}

}
```

Then you shouldn't need to manually call the method.

NachoSoto replied:
I see this is where I got it wrong, sorry! I should have realized you had `SKProductSubscriptionPeriod` as the parameter not `SubscriptionPeriod`.

NachoSoto commented on Tests/UnitTests/Purchasing/StoreKitAbstractions/StoreKitWorkaroundsTests.swift:
```diff
@@ -0,0 +1,70 @@
+
+            expect(normalized).to(
+                equal(expected),
+                description: "Expected \(period.debugDescription) to become \(expected.debugDescription)."
```
Then you can use the same `"Expected \(normalized) to become \(expected)"` here too.
